File: rust/ome/src/risk_engine.rs
use common::{Side, model::Balance};
use std::collections::HashMap;

const BTC_SCALE: i64 = 100_000_000;

pub struct RiskEngine {
    accounts: HashMap<u64, HashMap<i32, Balance>>,
    last_processed_seq_id: u64,
}

impl RiskEngine {
    pub fn new() -> Self {
        Self {
            accounts: HashMap::new(),
            last_processed_seq_id: 0,
        }
    }

    pub fn deposit(&mut self, user_id: u64, currency_id: i32, amount: i64) {
        let user_balances = self.accounts.entry(user_id).or_default();
        let balance = user_balances.entry(currency_id).or_default();
        balance.available += amount;
    }

    pub fn withdraw(&mut self, user_id: u64, currency_id: i32, amount: i64, seq_id: u64) -> bool {
        if seq_id != 0 && seq_id <= self.last_processed_seq_id {
            return false;
        }

        if let Some(user_balances) = self.accounts.get_mut(&user_id) {
            if let Some(balance) = user_balances.get_mut(&currency_id) {
                if balance.available >= amount {
                    balance.available -= amount;
                    if seq_id != 0 {
                        self.last_processed_seq_id = seq_id;
                    }
                    return true;
                }
            }
        }
        false
    }

    pub fn pre_check_order(&mut self, user_id: u64, side: Side, price: i64, qty: i64, seq_id: u64) -> bool {
        if seq_id != 0 && seq_id <= self.last_processed_seq_id {
            return false;
        }

        let currency_id = if side == Side::Buy { 2 } else { 1 };
        let required_amount = if side == Side::Buy {
            (price * qty) / BTC_SCALE
        } else {
            qty
        };

        let user_balances = self.accounts.entry(user_id).or_default();
        let balance = user_balances.entry(currency_id).or_default();
        
        if balance.available >= required_amount {
            balance.available -= required_amount;
            balance.locked += required_amount;
            if seq_id != 0 {
                self.last_processed_seq_id = seq_id;
            }
            return true;
        }
        false
    }

    pub fn on_trade(
        &mut self,
        maker_user_id: u64,
        taker_user_id: u64,
        taker_side: Side,
        price: i64,
        qty: i64,
    ) {
        let cost = (price * qty) / BTC_SCALE;

        if taker_side == Side::Buy {
            self.adjust_balance(taker_user_id, 2, 0, -cost);
            self.adjust_balance(taker_user_id, 1, qty, 0);
            self.adjust_balance(maker_user_id, 1, 0, -qty);
            self.adjust_balance(maker_user_id, 2, cost, 0);
        } else {
            self.adjust_balance(taker_user_id, 1, 0, -qty);
            self.adjust_balance(taker_user_id, 2, cost, 0);
            self.adjust_balance(maker_user_id, 2, 0, -cost);
            self.adjust_balance(maker_user_id, 1, qty, 0);
        }
    }

    pub fn on_cancel(&mut self, user_id: u64, side: Side, price: i64, leaves_qty: i64) {
        let currency_id = if side == Side::Buy { 2 } else { 1 };
        let unlock_amount = if side == Side::Buy {
            (price * leaves_qty) / BTC_SCALE
        } else {
            leaves_qty
        };

        self.adjust_balance(user_id, currency_id, unlock_amount, -unlock_amount);
    }

    fn adjust_balance(
        &mut self,
        user_id: u64,
        currency_id: i32,
        avail_delta: i64,
        lock_delta: i64,
    ) {
        let user_balances = self.accounts.entry(user_id).or_default();
        let balance = user_balances.entry(currency_id).or_default();
        
        if lock_delta < 0 {
            let to_deduct = -lock_delta;
            let actual_deduction = to_deduct.min(balance.locked);
            let remainder = to_deduct - actual_deduction;
            
            balance.locked -= actual_deduction;
            // If refunding (avail_delta > 0), only refund what was actually locked
            if avail_delta > 0 {
                balance.available += actual_deduction;
            } else {
                balance.available += avail_delta;
                balance.available -= remainder;
            }
        } else {
            balance.available += avail_delta;
            balance.locked += lock_delta;
        }
    }

    #[cfg(test)]
    pub fn get_balance(&self, user_id: u64, currency_id: i32) -> Option<&Balance> {
        self.accounts.get(&user_id)?.get(&currency_id)
    }
}

File: engines/jvm/src/main/kotlin/com/exchange/ome/RiskEngine.kt
package com.exchange.ome

import com.exchange.ome.model.Account
import com.exchange.sbe.Side
import com.exchange.sbe.OrderType
import org.eclipse.collections.impl.map.mutable.primitive.LongObjectHashMap

class RiskEngine {
    private val accounts = LongObjectHashMap<Account>()
    // Tracking last processed seqId per user for idempotency
    private val lastProcessedSeqId = org.eclipse.collections.impl.map.mutable.primitive.LongLongHashMap()

    fun getAccount(userId: Long): Account {
        var account = accounts.get(userId)
        if (account == null) {
            account = Account(userId)
            accounts.put(userId, account)
        }
        return account
    }

    private fun adjustBalance(userId: Long, currencyId: Int, availDelta: Long, lockDelta: Long) {
        val account = getAccount(userId)
        val balance = account.getBalance(currencyId)

        if (lockDelta < 0) {
            val toDeduct = -lockDelta
            val actualDeduction = Math.min(toDeduct, balance.locked)
            val remainder = toDeduct - actualDeduction
            
            balance.locked -= actualDeduction
            if (availDelta > 0) {
                // Profit scenario
                balance.available += availDelta
            } else {
                // Cost scenario: deduct remainder from available
                balance.available = Math.max(0, balance.available + availDelta - remainder)
            }
        } else {
            balance.available = Math.max(0, balance.available + availDelta)
            balance.locked = Math.max(0, balance.locked + lockDelta)
        }
    }

    /**
     * Pre-trade Risk Check (New Order)
     */
    fun preCheckOrder(userId: Long, symbolId: Int, side: Side, price: Long, qty: Long, seqId: Long, type: OrderType = OrderType.Limit): Boolean {
        if (lastProcessedSeqId.containsKey(userId) && lastProcessedSeqId.get(userId) >= seqId) {
            return false
        }

        val currencyId = if (side == Side.Buy) 2 else 1

        // Market Buy Orders (price=0) protection
        val effectivePrice = if (side == Side.Buy && price == 0L) {
            100_000_000L // 1.0 safety price
        } else {
            price
        }

        val requiredAmount = if (side == Side.Buy) {
            (effectivePrice * qty) / 100_000_000L
        } else {
            qty
        }

        val account = getAccount(userId)
        val balance = account.getBalance(currencyId)

        if (balance.available >= requiredAmount) {
            adjustBalance(userId, currencyId, -requiredAmount, requiredAmount)
            lastProcessedSeqId.put(userId, seqId)
            return true
        }
        return false
    }

    /**
     * Handle Execution Report (Trade Settlement)
     */
    fun onTrade(makerUserId: Long, takerUserId: Long, side: Side, price: Long, qty: Long) {
        val cost = (price * qty) / 100_000_000L
        val baseCurrency = 1 // BTC
        val quoteCurrency = 2 // KRW

        if (side == Side.Buy) { // Taker is Buying
            // Taker
            adjustBalance(takerUserId, quoteCurrency, 0, -cost)
            adjustBalance(takerUserId, baseCurrency, qty, 0)
            // Maker
            adjustBalance(makerUserId, baseCurrency, 0, -qty)
            adjustBalance(makerUserId, quoteCurrency, cost, 0)
        } else { // Taker is Selling
            // Taker
            adjustBalance(takerUserId, baseCurrency, 0, -qty)
            adjustBalance(takerUserId, quoteCurrency, cost, 0)
            // Maker
            adjustBalance(makerUserId, quoteCurrency, 0, -cost)
            adjustBalance(makerUserId, baseCurrency, qty, 0)
        }
    }
    
    fun onDeposit(userId: Long, currencyId: Int, amount: Long) {
        adjustBalance(userId, currencyId, amount, 0)
    }

    fun onWithdrawRequest(userId: Long, currencyId: Int, amount: Long, seqId: Long = 0): Boolean {
        if (seqId != 0L && lastProcessedSeqId.containsKey(userId) && lastProcessedSeqId.get(userId) >= seqId) {
            return false
        }

        val balance = getAccount(userId).getBalance(currencyId)
        if (balance.available >= amount) {
            adjustBalance(userId, currencyId, -amount, amount)
            if (seqId != 0L) lastProcessedSeqId.put(userId, seqId)
            return true
        }
        return false
    }
    
    fun onCancel(orderId: Long, side: Side, price: Long, qty: Long, userId: Long) {
        val isBuy = (side == Side.Buy)
        val currencyId = if (isBuy) 2 else 1
        val unlockAmount = if (isBuy) (price * qty) / 100_000_000L else qty
        
        adjustBalance(userId, currencyId, unlockAmount, -unlockAmount)
        println("Risk: Order $orderId Cancelled. Refunded $unlockAmount to User $userId")
    }
    
    fun deposit(userId: Long, currencyId: Int, amount: Long) {
        onDeposit(userId, currencyId, amount)
    }
}
